File: Safe/src/main/java/org/openintents/safe/RestoreFirstTimeActivity.java
package org.openintents.safe;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.TextView;

import org.openintents.util.VersionUtils;

public class RestoreFirstTimeActivity extends AppCompatActivity {
    private Button restore;
    private Button chooseFile;
    private Button cancel;
    private String path;

    private static final int REQUEST_RESTORE = 0;
    private static final int REQUEST_CHOOSE_FILE = 1;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.restore_first_time);
        path = PreferenceActivity.getBackupPath(this);
        restore = (Button) findViewById(R.id.restore);
        chooseFile = (Button) findViewById(R.id.choose_file);
        cancel = (Button) findViewById(R.id.cancel);

        ((TextView) findViewById(R.id.filename)).setText(path);
        restore.setOnClickListener(
                new OnClickListener() {
                    public void onClick(View v) {
                        Intent i = new Intent(RestoreFirstTimeActivity.this, Restore.class);
                        i.putExtra(Restore.KEY_FILE_PATH, path);
                        i.putExtra(Restore.KEY_FIRST_TIME, true);
                        startActivityForResult(i, REQUEST_RESTORE);
                    }
                }
        );
        chooseFile.setOnClickListener(
                new OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
                        intent.addCategory(Intent.CATEGORY_OPENABLE);
                        intent.setType("*/*");
                        startActivityForResult(intent, REQUEST_CHOOSE_FILE);
                    }
                }
        );
        cancel.setOnClickListener(
                new OnClickListener() {
                    public void onClick(View v) {
                        setResult(RESULT_CANCELED);
                        finish();
                    }
                }
        );

        /* Copied from AskPassword.java - normalInit() */
        TextView header = (TextView) findViewById(R.id.entry_header);
        String version = VersionUtils.getVersionNumber(this);
        String appName = VersionUtils.getApplicationName(this);
        String head = appName + " " + version;
        header.setText(head);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case REQUEST_RESTORE:
                setResult(resultCode);
                finish();
                break;
            case REQUEST_CHOOSE_FILE:
                if (resultCode == RESULT_OK && data != null) {
                    path = data.getData().toString();
                    ((TextView) findViewById(R.id.filename)).setText(path);
                }
                break;
        }
    }
}

File: Safe/src/main/java/io/github/cfsima/modernsafe/RestoreFirstTimeActivity.kt
package io.github.cfsima.modernsafe

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import io.github.cfsima.modernsafe.ui.theme.OISafeTheme
import java.io.File
import io.github.cfsima.modernsafe.util.VersionUtils

class RestoreFirstTimeActivity : AppCompatActivity() {

    private val restoreLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        setResult(result.resultCode)
        finish()
    }

    private val chooseFileLauncher = registerForActivityResult(ActivityResultContracts.OpenDocument()) { uri: Uri? ->
        uri?.let {
            val intent = Intent(this, Restore::class.java)
            intent.data = it
            intent.putExtra(Restore.KEY_FIRST_TIME, true)
            restoreLauncher.launch(intent)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val defaultPath = Settings.getBackupPath(this)
        val fileExists = File(defaultPath).exists()

        val appName = VersionUtils.getApplicationName(this)
        val version = VersionUtils.getVersionNumber(this)
        val headerText = "$appName $version"

        setContent {
            OISafeTheme {
                Surface(modifier = Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background) {
                    RestoreFirstTimeScreen(
                        headerText = headerText,
                        filePath = defaultPath,
                        fileExists = fileExists,
                        onRestore = {
                            val intent = Intent(this, Restore::class.java)
                            intent.putExtra(Restore.KEY_FILE_PATH, defaultPath)
                            intent.putExtra(Restore.KEY_FIRST_TIME, true)
                            restoreLauncher.launch(intent)
                        },
                        onChooseFile = {
                             chooseFileLauncher.launch(arrayOf("*/*"))
                        },
                        onCancel = {
                            setResult(Activity.RESULT_CANCELED)
                            finish()
                        }
                    )
                }
            }
        }
    }
}

@Composable
fun RestoreFirstTimeScreen(
    headerText: String,
    filePath: String,
    fileExists: Boolean,
    onRestore: () -> Unit,
    onChooseFile: () -> Unit,
    onCancel: () -> Unit
) {
    Column(
        modifier = Modifier.fillMaxSize().padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Text(text = headerText, style = MaterialTheme.typography.titleLarge)
        Spacer(modifier = Modifier.height(32.dp))

        Text(text = stringResource(R.string.filename))
        Text(text = filePath)

        if (!fileExists) {
            Spacer(modifier = Modifier.height(8.dp))
            Text(text = stringResource(R.string.restore_no_file), color = MaterialTheme.colorScheme.error)
        }

        Spacer(modifier = Modifier.height(32.dp))

        Button(onClick = onRestore, enabled = fileExists) {
            Text(stringResource(R.string.restore))
        }

        Spacer(modifier = Modifier.height(16.dp))

        Button(onClick = onChooseFile) {
            Text(stringResource(R.string.restore_select_file))
        }

        Spacer(modifier = Modifier.height(16.dp))

        OutlinedButton(onClick = onCancel) {
            Text(stringResource(android.R.string.cancel))
        }
    }
}
